File: core/examples/companion-example.ts
/**
 * Example: Frontier Companion API (CAPI)
 *
 * Prerequisites:
 * 1. Register an application at https://user.frontierstore.net/developer
 * 2. Get your Client ID and set up a redirect URI
 *
 * Flow:
 * 1. Generate PKCE challenge
 * 2. Open auth URL in browser for user to approve
 * 3. Exchange the auth code for tokens
 * 4. Make API requests
 *
 * Tokens are persisted to a JSON file so re-authorization is only needed
 * when the refresh token expires or is revoked.
 */

import {
  CompanionClient,
  FrontierAuth,
  generateCodeChallenge,
  generateCodeVerifier,
} from "@elite-dangerous-sdk/companion";
import * as fs from "node:fs";
import * as path from "node:path";

// === Token persistence ===
const TOKEN_FILE = path.join(
  process.env.XDG_CONFIG_HOME || path.join(process.env.HOME || process.env.USERPROFILE || ".", ".config"),
  "elite-dangerous-sdk",
  "companion-tokens.json",
);

function _saveTokens(tokens: { accessToken: string; refreshToken: string; expiresAt: number }) {
  const dir = path.dirname(TOKEN_FILE);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(TOKEN_FILE, JSON.stringify(tokens, null, 2), "utf-8");
  console.log(`Tokens saved to ${TOKEN_FILE}`);
}

function _loadTokens(): { accessToken: string; refreshToken: string; expiresAt: number } | null {
  try {
    const raw = fs.readFileSync(TOKEN_FILE, "utf-8");
    return JSON.parse(raw);
  } catch {
    return null;
  }
}

// === Step 1: Configure auth ===
const CLIENT_ID = "your-client-id-here";
const REDIRECT_URI = "http://localhost:8080/callback";
const APP_NAME = "MyEliteTool";
const APP_VERSION = "1.0.0";

const auth = new FrontierAuth({
  clientId: CLIENT_ID,
  redirectUri: REDIRECT_URI,
  appName: APP_NAME,
  appVersion: APP_VERSION,
});

// Try to restore a previous session
const saved = _loadTokens();
if (saved) {
  auth.setTokens(saved);
  console.log("Restored saved session.");
}

// === Step 2: Generate PKCE and get auth URL ===
async function _startAuthFlow() {
  const codeVerifier = generateCodeVerifier();
  const codeChallenge = await generateCodeChallenge(codeVerifier);
  const state = crypto.randomUUID();

  const authUrl = auth.buildAuthUrl(codeChallenge, state);
  console.log("Visit this URL to authorize:");
  console.log(authUrl);

  // In a real app, you'd open the browser and wait for the redirect
  // The redirect will include ?code=AUTH_CODE&state=STATE
  return { codeVerifier, state };
}

// === Step 3: Exchange code for tokens ===
async function _handleCallback(
  authCode: string,
  codeVerifier: string,
  state: string,
  expectedState: string,
) {
  const tokens = await auth.exchangeCode(
    authCode,
    codeVerifier,
    state,
    expectedState,
  );
  console.log(
    `Access token expires at: ${new Date(tokens.expiresAt).toISOString()}`,
  );

  _saveTokens({ accessToken: tokens.accessToken, refreshToken: tokens.refreshToken, expiresAt: tokens.expiresAt });
}

// === Step 4: Use the API ===
async function _useApi() {
  const client = new CompanionClient({ auth });

  // Check if commander is docked before market reads
  const docked = await client.isDocked();
  console.log(`Docked: ${docked}`);

  if (docked) {
    // Get profile
    const profile = await client.getProfile();
    console.log(`Commander: ${profile.commander?.name}`);
    console.log(`Credits: ${profile.commander?.credits}`);

    // Get market data
    const market = await client.getMarket();
    console.log(`Station: ${market.lastStarport?.name}`);
    console.log(`Market items: ${market.market?.items?.length ?? 0}`);

    // Get shipyard
    const shipyard = await client.getShipyard();
    console.log(`Ships available: ${shipyard.ships?.length ?? 0}`);
    console.log(`Modules available: ${shipyard.modules?.length ?? 0}`);
  }

  // Get fleet carrier info
  const fc = await client.getFleetCarrier();
  if (fc.name) {
    console.log(`Fleet Carrier: ${fc.name}`);
    console.log(`Balance: ${fc.currentBalance}`);
    console.log(`Fuel: ${fc.fuelLevel}`);
  }

  // Get community goals
  const cgs = await client.getCommunityGoals();
  console.log(`Active goals: ${cgs.communitygoals?.length ?? 0}`);
}

// === Step 5: Handle token refresh (automatic) ===
async function _demonstrateRefresh() {
  // The SDK automatically refreshes when tokens are close to expiry
  // You can also manually check:
  console.log(`Authenticated: ${auth.isAuthenticated()}`);

  const stored = auth.getStoredTokens();
  if (stored) {
    console.log(
      `Current access token: ${stored.accessToken.substring(0, 20)}...`,
    );
    console.log(`Expires at: ${new Date(stored.expiresAt).toISOString()}`);
  }

  // Tokens are reloaded from file on next startup (see top of file)
}

// To clear auth and delete saved tokens:
// auth.clearTokens();
// fs.unlinkSync(TOKEN_FILE);

File: python/examples/companion_example.py
"""
Example: Frontier Companion API (CAPI)

Prerequisites:
1. Register an application at https://user.frontierstore.net/developer
2. Get your Client ID and set up a redirect URI

Flow:
1. Generate PKCE challenge
2. Open auth URL in browser for user to approve
3. Exchange the auth code for tokens
4. Make API requests

Tokens are persisted to a JSON file so re-authorization is only needed
when the refresh token expires or is revoked.
"""

import json
import os
import secrets
from pathlib import Path

from elite_dangerous_sdk.companion import (
    CompanionClient,
    FrontierAuth,
    generate_code_challenge,
    generate_code_verifier,
)

# === Token persistence ===
CONFIG_DIR = Path(os.environ.get("XDG_CONFIG_HOME", Path.home() / ".config"))
TOKEN_FILE = CONFIG_DIR / "elite-dangerous-sdk" / "companion-tokens.json"


def _save_tokens(tokens):
    TOKEN_FILE.parent.mkdir(parents=True, exist_ok=True)
    with open(TOKEN_FILE, "w") as f:
        json.dump({
            "access_token": tokens.access_token,
            "refresh_token": tokens.refresh_token,
            "expires_at": tokens.expires_at,
        }, f, indent=2)
    print(f"Tokens saved to {TOKEN_FILE}")


def _load_tokens():
    try:
        with open(TOKEN_FILE) as f:
            return json.load(f)
    except (FileNotFoundError, json.JSONDecodeError):
        return None


# === Step 1: Configure auth ===
CLIENT_ID = "your-client-id-here"
REDIRECT_URI = "http://localhost:8080/callback"
APP_NAME = "MyEliteTool"
APP_VERSION = "1.0.0"

auth = FrontierAuth(
    client_id=CLIENT_ID,
    redirect_uri=REDIRECT_URI,
    app_name=APP_NAME,
    app_version=APP_VERSION,
)

# Try to restore a previous session
saved = _load_tokens()
if saved:
    from elite_dangerous_sdk.companion import FrontierTokens
    auth.tokens = FrontierTokens(
        access_token=saved["access_token"],
        refresh_token=saved["refresh_token"],
        expires_at=saved["expires_at"],
    )
    print("Restored saved session.")


# === Step 2: Generate PKCE and get auth URL ===
def start_auth_flow():
    code_verifier = generate_code_verifier()
    code_challenge = generate_code_challenge(code_verifier)
    state = secrets.token_urlsafe(16)

    auth_url = auth.build_auth_url(code_challenge, state)
    print("Visit this URL to authorize:")
    print(auth_url)

    return code_verifier, state


# === Step 3: Exchange code for tokens ===
def handle_callback(auth_code, code_verifier, state, expected_state):
    tokens = auth.exchange_code(auth_code, code_verifier, state, expected_state)
    print(f"Access token expires at: {tokens.expires_at}")

    _save_tokens(tokens)


# === Step 4: Use the API ===
def use_api():
    client = CompanionClient(auth)

    # Check if commander is docked before market reads
    docked = client.is_docked()
    print(f"Docked: {docked}")

    if docked:
        profile = client.get_profile()
        print(f"Commander: {profile.get('commander', {}).get('name')}")
        print(f"Credits: {profile.get('commander', {}).get('credits')}")

        market = client.get_market()
        station = market.get("lastStarport", {})
        print(f"Station: {station.get('name')}")
        items = market.get("market", {}).get("items", [])
        print(f"Market items: {len(items)}")

        shipyard = client.get_shipyard()
        ships = shipyard.get("ships", []) or shipyard.get("shipyard", {}).get("ships", [])
        print(f"Ships available: {len(ships)}")

    fc = client.get_fleet_carrier()
    if fc.get("name"):
        print(f"Fleet Carrier: {fc['name']}")
        print(f"Balance: {fc.get('currentBalance')}")
        print(f"Fuel: {fc.get('fuelLevel')}")

    cgs = client.get_community_goals()
    goals = cgs.get("communitygoals", [])
    print(f"Active goals: {len(goals)}")


# === Step 5: Handle token refresh (automatic) ===
def demonstrate_refresh():
    print(f"Authenticated: {auth.tokens is not None}")
    if auth.tokens:
        print(f"Current access token: {auth.tokens.access_token[:20]}...")
        print(f"Expires at: {auth.tokens.expires_at}")


# To clear auth and delete saved tokens:
# auth.tokens = None
# TOKEN_FILE.unlink()
